Feature Programming Exercise: Allow tutors to respond to complaints

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Students are already able to create complaints and more feedback requests, but the tutors can't respond to them. This PR aims to add complaints inside the `Update Result` dialog.

### Description
<!-- Describe your changes in detail -->
Tutors can respond to complaint from both 'List of Complaints' and 'Programming Exercise Dashboard' pages. The complaint is located inside the 'Update Manual Result' dialog.

Additionally, this feature brings some code duplication to the server, which will be solved in follow-up PRs.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Select 'Tutor Course Dashboard'
4. Select 'Exercise Dashboard' on Programming Exercise
5. Click 'Update Result' in Complaints list

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
#### Resolving Complaint
![Peek 2019-11-13 21-45](https://user-images.githubusercontent.com/17262376/68802911-23c51600-065f-11ea-8b16-198f5cb61416.gif)
#### List of Complaints
![Peek 2019-11-13 21-49](https://user-images.githubusercontent.com/17262376/68803166-84545300-065f-11ea-9614-431c82018062.gif)

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:
```diff
@@ -0,0 +1,95 @@
+
+        if (result.getParticipation() != null && result.getParticipation() instanceof StudentParticipation
+                && !authCheckService.isAtLeastInstructorForExercise(programmingExercise, user)) {
+            ((StudentParticipation) result.getParticipation()).setStudent(null);
```
```suggestion
            ((StudentParticipation) result.getParticipation()).filterSensitiveInformation();
```

maxr96 replied:
Done

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:
```diff
@@ -0,0 +1,95 @@
+        }
+
+        if (result.getParticipation() != null && result.getParticipation() instanceof StudentParticipation
+                && !authCheckService.isAtLeastInstructorForExercise(programmingExercise, user)) {
```
You might be able to replace this Block by using
```java
programmingSubmissionService.hideDetails(programmingSubmission);
```

Can you check if that is a possibility here, as I am not quite sure.

maxr96 replied:
No. This is used for submission and we never use it in this case.

jpbernius commented on src/main/webapp/app/programming-assessment/manual-result/programming-assessment-manual-result-dialog.component.ts:
```diff
@@ -61,12 +75,22 @@ export class ProgrammingAssessmentManualResultDialogComponent implements OnInit
-        this.participation = this.result.participation! as StudentParticipation;
+        if (this.result.hasComplaint) {
+            this.getComplaint(this.result.id);
+            this.getParticipation();
```
Can you explain to me, why we need this here? I cannot follow why
```ts
this.participation = this.result.participation! as StudentParticipation;
```
does not work in this case.

maxr96 replied:
What I discovered is that in case that we have complaint, we can't get from it's participation student's name. In order to get it I explicitly ask again for the participation

jpbernius replied:
Do we really want to get the students name? Or should it be hidden?

maxr96 replied:
I am not sure here. Maybe @krusche can give more insight. However, inside the manual result dialog we explicitly show student's name

jpbernius commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.html:
```diff
@@ -310,13 +309,24 @@ <h2>
+                                            (click)="
+                                                exercise.type === ExerciseType_PROGRAMMING
+                                                    ? openManualResultDialog(complaint.result.participation.id, complaint.result)
+                                                    : openAssessmentEditor(complaint.result.submission.id, false)
```
I am wondering if we can move this as one method into TypeScript and reuse this four times in the HTML.

(Also, see comment in `tutor-exercise-dashboard.component.ts` on change in parameters for `openManualResultDialog`.)

maxr96 replied:
The idea was to not parameters passed into the `openAssessmentEditor` since it is used like 7 times. But I can also do that.

jpbernius replied:
I am wondering if we should add a method like this:

```typescript
handleComplaint(complaint: Complaint) {
  if (this.exercise.type === ExerciseType.PROGRAMMING) {
    this.openManualResultDialog(complaint.result);
  } else {
    this.openAssessmentEditor(complaint.result.submission!.id, false)
  }
}
```

and thereby reduce the HTML to this:

```html
<button
  class="btn btn-primary btn-sm"
  (click)="handleComplaint(complaint)"
>
  ...
</button>
```
OR
```html
<button
  class="btn btn-primary btn-sm"
  (click)="handleComplaint(moreFeedbackRequest)"
>
  ...
</button>
```

maxr96 replied:
Great idea! Added it and called it `viewComplaint`

jpbernius commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts:
```diff
@@ -366,6 +370,19 @@ export class TutorExerciseDashboardComponent implements OnInit {
         this.router.navigate([route]);
     }
 
+    openManualResultDialog(participationId: number, result: Result) {
```
```suggestion
    openManualResultDialog(result: Result) {
        const participationId: number = result.participation.id; 
```

maxr96 replied:
Missed it. Thanks.

maxr96 commented:
> Thanks for this pull request @maxr96!
> I have some ideas and questions:

Thank you for the comments @jpbernius . I applied some of the suggested changes and left some comments.

ungaralex commented on src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:
```diff
@@ -78,7 +78,9 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
         // Update the result that was complained about with the new feedback
         originalResult.updateAllFeedbackItems(assessmentUpdate.getFeedbacks());
-        originalResult.evaluateFeedback(exercise.getMaxScore());
+        if (!(exercise instanceof ProgrammingExercise)) {
```
WIll you also refactor this service in the code duplication PR? Because I would suggest to have some inheritance here in order to avoid `instanceof` checks.

maxr96 replied:
Yes, this will be fixed in the code duplication PR

maxr96 commented on src/main/webapp/app/programming-assessment/manual-result/programming-assessment-manual-result.service.ts:
```diff
@@ -34,6 +34,18 @@ export class ProgrammingAssessmentManualResultService {
             .map((res: EntityResponseType) => this.resultService.convertDateFromServer(res));
     }
 
+    updateAfterComplaint(feedbacks: Feedback[], complaintResponse: ComplaintResponse, result: Result, submissionId: number): Observable<Result> {
```
Added

maxr96 commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts:
```diff
@@ -366,6 +370,27 @@ export class TutorExerciseDashboardComponent implements OnInit {
+            this.openAssessmentEditor(complaint.result.participation!.id, false);
+        }
+    }
+
```
Added

maxr96 commented:
> I have tested your feature and found some issues within the implementation. Can you please have a look at them? Thank you :)
> 
>     * When a tutor navigates to the List of complaints page, then accepted complaints cannot be opened by clicking `Show the complaint`. Also, the submission ID is missing (see screenshot).
>       ![image](https://user-images.githubusercontent.com/33764166/68983848-48f78700-080d-11ea-8619-78d41e0b62ea.png)
> 
>     * When the tutor navigates to the List of complaints and opens the complaint for his/her own assessment, then it looks like he/she can update the assessment or reject the complaint (which is not possible -> only errors are thrown).  Wouldn't it be better to disable the buttons `update assessment` and `reject complaint` in this case and replace `cancel` and `save` with `close`?
> 
>     * Moreover a REST call for participations is made, e.g. `https://artemistest.ase.in.tum.de/api/participations/156496` where the student information is visible to the tutor. Shouldn't this information be hidden?
>       ![image](https://user-images.githubusercontent.com/33764166/68984023-6842e400-080e-11ea-85c3-b4e57cd2f9d6.png)

@lovaiible Thank you for checking my PR and reporting about the issues. 
I was not able to reproduce the first issue, for me it all worked good. This also seems to be related to how the complaints are created and not the programming exercise in particular. I guess it can be tackled in a separate bugfix.
I fixed the second issue, turned out to be I was getting the assessor too early. Now if the assessor and current user is the same it will only see the text of complaint, but can't edit it.
Regarding the last issue, the previous implementation was showing the student name. Now we got rid of it and don't show it anymore. The call to participation service was deleted for updating the result of complaint.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:
```diff
@@ -1196,4 +1196,16 @@ public long countSubmissionsToAssessByExerciseId(Long exerciseId) {
+     * @param id the id of the entity
+     * @return the entity
+     */
+    @Transactional(readOnly = true)
```
remove Transactional here!

maxr96 replied:
Removed

krusche replied:
It is still here?! 👎

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:
```diff
@@ -0,0 +1,41 @@
+     * @param assessmentUpdate the assessment update
+     * @return the updated Result
+     */
+    @Transactional
```
remove Transactional here and make sure this works without Transactional!

maxr96 replied:
Removed!

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/ProgrammingAssessmentUpdate.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * A helper object that aggregates feedback list, complaint response, score, result string and completion date which are used to update an assessment after a complaint.
+ */
+public class ProgrammingAssessmentUpdate extends AssessmentUpdate {
```
Why do we need this?

maxr96 replied:
This is similar to how we do it with other types of exercises, but in case of programming exercise tutor can also change result string, completion date and score. But again I am not sure if we should allow to change manual result infinitely as we have it now

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    private boolean areManualResultsAllowed(final Exercise exerciseToBeChecked) {
+        // Only allow manual results for programming exercises if option was enabled and due dates have passed
```
I think this method already exists somewhere else. Please reuse code and do not simply copy it!

maxr96 replied:
Yes it is used inside the `ResultResource`. I would like to move it into the `ProgrammingAssessmentService`, but I don't want to inject it this service in the `ResultResource`. 
We actually need to remove all the methods specific for programming exercise from the `ResultResource` to this file, then we can easily reuse this method

maxr96 replied:
Moved it to `ProgrammingExercise` domain

krusche commented:
Why are there no client tests? Please add them!

krusche commented:
Also make sure that you do not use **Transactional** any more. This is bad design, we should put this into our design guidelines that Transactional should be avoided if it is not absolutely necessary!!!

jpbernius commented on src/main/webapp/app/list-of-complaints/list-of-complaints.component.ts:
```diff
@@ -9,6 +9,8 @@ import { Observable } from 'rxjs';
 import { ExerciseType } from 'app/entities/exercise';
 import * as moment from 'moment';
 import { StudentParticipation } from 'app/entities/participation';
+import { ProgrammingAssessmentManualResultDialogComponent } from 'app/programming-assessment/manual-result';
```
Please import component directly

maxr96 replied:
Changed all imports

jpbernius commented on src/main/webapp/app/programming-assessment/manual-result/programming-assessment-manual-result-dialog.component.ts:
```diff
@@ -1,18 +1,21 @@
 import { DatePipe } from '@angular/common';
 import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
-import { Result } from '../../entities/result/result.model';
+import { Result } from 'app/entities/result';
```
```suggestion
import { Result } from 'app/entities/result/result.model';
```

jpbernius commented on src/main/webapp/app/programming-assessment/manual-result/programming-assessment-manual-result-dialog.component.ts:
```diff
@@ -1,18 +1,21 @@
+import { catchError, tap, filter } from 'rxjs/operators';
 import { ProgrammingAssessmentManualResultService } from 'app/programming-assessment/manual-result/programming-assessment-manual-result.service';
 import { SCORE_PATTERN } from 'app/app.constants';
+import { Complaint, ComplaintService, ComplaintType } from 'app/entities/complaint';
```
Please do not import from barrel files.

maxr96 replied:
Changed

jpbernius commented on src/main/webapp/app/programming-assessment/manual-result/programming-assessment-manual-result.service.ts:
```diff
@@ -4,6 +4,8 @@ import { SERVER_API_URL } from 'app/app.constants';
 import { HttpClient } from '@angular/common/http';
 import * as moment from 'moment';
+import { Feedback } from 'app/entities/feedback';
+import { ComplaintResponse } from 'app/entities/complaint-response';
```
Please import directly

jpbernius commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts:
```diff
@@ -26,6 +26,9 @@ import { FileUploadSubmissionService } from 'app/entities/file-upload-submission
 import { ProgrammingExercise } from 'app/entities/programming-exercise';
 import { ProgrammingSubmissionService } from 'app/programming-submission';
+import { Result } from 'app/entities/result';
+import { ProgrammingAssessmentManualResultDialogComponent } from 'app/programming-assessment/manual-result';
```
Please import directly

HanyaElhashemy commented:
@krusche I think the logic here is not complete as well since the assessment due date is also missing (same situation as with the file upload exercises we discussed yesterday). But maybe this should be done in a separate PR?

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:
```diff
@@ -0,0 +1,79 @@
+        ProgrammingSubmission programmingSubmission = programmingSubmissionService.findOneWithEagerResultAndFeedback(submissionId);
+        Participation participation = programmingSubmission.getParticipation();
+        long exerciseId = participation.getExercise().getId();
+        ProgrammingExercise programmingExercise = programmingExerciseService.findOne(exerciseId);
```
You are right we don't have to get exercise again

jpbernius replied:
This is not resolved!

maxr96 replied:
You are right. Changed the query and code

jpbernius commented:
I found another problem: I think the Programming Exercises Score Pages does not reflect changes made because of a complaint.

maxr96 commented:
> I found another problem: I think the Programming Exercises Score Pages does not reflect changes made because of a complaint.

@jpbernius Do you mean that it throws exception? What Sentry shows me is that it is happening after we check for file upload, modeling and text exercise, so programming exercise should not affect it. I don't understand yet why this happens.

beatrisbu commented:
Nice feature! 

I marked the problems that I found with red rectangulars on the screenshots below.

(I am not sure if this one is a part of this PR) The number of my Assessments as well as the number of the Complaints remains 0 when it should be 1:
![step 1](https://user-images.githubusercontent.com/33368508/69272355-cee14c80-0bd6-11ea-96e0-ee048df34398.png)

I don't have/see/get a Complaint response window:
![step 2](https://user-images.githubusercontent.com/33368508/69272368-d3a60080-0bd6-11ea-90a4-0d291fb24ef5.png)

Also one note (which might as well not be a part of this PR) - it takes 5+ seconds in order for the numbers on the right side on the 1st screenshot to load, which is quite a long time.

maxr96 commented:
> Hi @maxr96, somehow after your latest changes I haven't been able to respond to a complaint:
> Please log in as `test user 6` and go to the page https://artemistest.ase.in.tum.de/#/course/19/exercise/1223/tutor-dashboard for example.
> ![image](https://user-images.githubusercontent.com/33764166/69272434-eb7d8480-0bd6-11ea-84f6-3ab9392415ce.png)
> 
> Moreover if I don't add an assessment text, the console throws the following error. It would be better to show an alert or to allow no text assessments:
> ![image](https://user-images.githubusercontent.com/33764166/69272668-42835980-0bd7-11ea-876e-c1b3462b4bf6.png)
> 
> And another feedback: I don't think the date picker in the complaint modal changes anything, but it would be better to keep the date-picker disabled here.
> ![image](https://user-images.githubusercontent.com/33764166/69272990-dfde8d80-0bd7-11ea-9d5d-035ea867a85c.png)
> 
> But if you could disable the date-picker in the manual assessment modal as well, this would be great:
> The reason is, if manual results are allowed, then the tutor can change the submission date of a late submission and so that the manual assessment can be counted to the student's score.
> ![image](https://user-images.githubusercontent.com/33764166/69272954-c8070980-0bd7-11ea-8c9e-3f388d8179b4.png)

Thank you for checking my PR @lovaiible 
1. The due date hasn't passed yet, that is why you are not able to resolve the complaint.
2. Added a warning, it should not throw exception anymore.
3. Good idea with the date-picker, I disabled it completely.

krusche commented:
@ungaralex and I fixed most of the important issues so that this should work much more reliable.
By doing that we also fixed some other issue related to this feature in general.

It would be great if everyone can again test this PR on the test server, so that we can merge and release it tonight!

HanyaElhashemy commented:
The main functionality of this feature works and I can respond to complaints!
Some issues mentioned above still exist and others I found now but I think they can be handled in a follow-up PR.
Here are my findings that could be handled in a different PR:
- Instructors can't see complaints that have been evaluated by a tutor anymore
expected behavior here: instructors should be able to see all 2 complaints
![image](https://user-images.githubusercontent.com/50635005/69349557-3f8c7580-0c78-11ea-9ada-00c5a2974cbe.png)
![image](https://user-images.githubusercontent.com/50635005/69349579-4a470a80-0c78-11ea-88d3-1233ee579613.png)
- Assessors can see complaints about their own assessment and click on evaluate complaint -> disable button
here I am logged with the account of the assessor but can still see the complaints about my own assessment and even click on evaluate:
![image](https://user-images.githubusercontent.com/50635005/69349699-81b5b700-0c78-11ea-99aa-9f21da244c0f.png)
![image](https://user-images.githubusercontent.com/50635005/69349792-af9afb80-0c78-11ea-97ca-6b79fc4793f1.png)
- programming exercise does not show in the instructors dashboard
- complaints about your assessment do not reflect the correct number, shows 0 but when we click on 0 it shows one complaint (mentioned above)

beatrisbu commented:
I can still observe the behavior I was observing before. 
One other thing that I noticed is that the numbers on the tutor exercise dashboard are also not updated (+some console errors):

![tutor exercise dashboard](https://user-images.githubusercontent.com/33368508/69373148-79bf3c80-0ca3-11ea-96e9-6889b46f7f6a.png)

krusche commented:
Thanks to everyone for testing. As we still have too many issues in this PR, we move it to the next release 3.9.4

maxr96 commented:
> why did you remove this? I added this on purpose!

I updated it to 'View Result', because it is used only for showing results with complaints to the assessor. So the tutor can't either resolve the complaint or update the result, but only view both result and complaint.

maxr96 commented:
> The main functionality of this feature works and I can respond to complaints!
> Some issues mentioned above still exist and others I found now but I think they can be handled in a follow-up PR.
> Here are my findings that could be handled in a different PR:
> 
>     * Assessors can see complaints about their own assessment and click on evaluate complaint -> disable button
>       here I am logged with the account of the assessor but can still see the complaints about my own assessment and even click on evaluate:
>       ![image](https://user-images.githubusercontent.com/50635005/69349699-81b5b700-0c78-11ea-99aa-9f21da244c0f.png)
>       ![image](https://user-images.githubusercontent.com/50635005/69349792-af9afb80-0c78-11ea-97ca-6b79fc4793f1.png)
> 
>     * programming exercise does not show in the instructors dashboard
> 
>     * complaints about your assessment do not reflect the correct number, shows 0 but when we click on 0 it shows one complaint (mentioned above)

Thank you for the review @HanyaElhashemy . Regarding the issues you mentioned:
- The assessor indeed can see the complaint and this is also how it works for other exercise type. But I agree that the button title was misleading, so I changed it to 'View Result'
- Yes they don't show up there. I don't know when do we plan to add programming exercises to the instructor dashboard, but it is out of scope of this PR
- It is actually an intended behavior. In `Complaints about your assessments` we count only those complaints that were accepted. I believe this needs more explanation to the user or maybe we should rework it. I think that is a good point for discussion and should be changed in another PR

maxr96 commented:
> Hi @maxr96, I tested the new changes on the test server and found some other issues:
> 
>     * The number of complaints is still wrongly displayed. It is only correctly calculated when the complaint has been addressed. (In my screenshot, it should be 2 complaints. When I navigate to the list of complaints, two complaints are correctly displayed in the table).
>       ![image](https://user-images.githubusercontent.com/33764166/69352421-3651d780-0c7d-11ea-923b-c2217fc7cae8.png)
> 
>     * When accepting/rejecting complaints I receive a 404 error, see screenshots:
>       ![image](https://user-images.githubusercontent.com/33764166/69352573-6ef1b100-0c7d-11ea-98cf-9b25d8755437.png)
>       ![image](https://user-images.githubusercontent.com/33764166/69352583-744efb80-0c7d-11ea-9567-15ee58f4c3d2.png)
> 
>     * Rejecting complaints without entering a response text leads to the error, that the original tutor is replaced with the tutor who did the complaint assessment. And the complaint has still the status `no reply yet` (see screenshot: tutor who assessed the submission and received the complaint was test user 7, but was changed to test user 8 who assessed the complaint)
>       ![image](https://user-images.githubusercontent.com/33764166/69352671-9cd6f580-0c7d-11ea-8aa1-ba674dc5598d.png)
> 
>     * Afterwards test user 7 can assess his own complaint because the reviewer is wrongly saved as test user 8
> 
>     * After the complaints have been resolved, the buttons in the student submission table looks like this:
>       ![image](https://user-images.githubusercontent.com/33764166/69353009-1ff84b80-0c7e-11ea-8ee8-200fca7f23f1.png)
>       I think the label `evaluate the complaint` is misleading because the complaint cannot be evaluated here. It would be better to call it `view assessment` or `view manual result` instead imo.

Thank you for testing it @lovaiible It took me some time to fix the issue with result being assigned to the current user after resolving the complaint. Awesome that you spotted it, before it can affect real users :+1:  Regarding your feedback:
* Wrong numbers, as I already mentioned are right, since they only count accepted complaints. Will be solved in another PR.
* This is fixed now. We don't call the endpoint on each complaint response anymore
* Next two points are fixed now
* Good suggestion. I added View Manual Result as a title button for this use case

maxr96 commented:
> @maxr96 please make a note for your next refactoring PR that we are not using the same component for responding to complaints in the tutor dashboard. There is the `jhi-programming-assessment-manual-result` component referenced in the tutor dashboard template and a callback to a method `viewComplaint` in the complaints section for tutors. In the end, both methods create the same modal `ProgrammingAssessmentManualResultDialogComponent`. We either should have a separate modal for complaints or just use the same components. As of now, the used buttons are inconsistent depending on who is logged in (tutor vs. instructor).

@ungaralex Indeed it should be changed in the next PR. This will also include moving manual result handling to a separate page.

ungaralex commented on src/main/webapp/app/programming-assessment/manual-result/programming-assessment-manual-result.service.ts:
```diff
@@ -32,6 +34,24 @@ export class ProgrammingAssessmentManualResultService {
     }
 
+    /**
+     * Send request to update the assessment after the complaint
```
This is still not a good documentation. Please just describe very briefly what "updating" entails. What happens to the assessment? What exactly gets updated? Does it get updated all the time, or are there certain conditions? 
Also the parameters: I know that a feedback array is a list of feedback items, but how does this relate to the other parameters or the whole method? 

We are all guilty of often writing redundant documentation, but it would be great if we started trying to put ourselves in the shoes of a new developer in the project.

ungaralex replied:
This is especially for the service methods since they are shared between multiple components and will be often reused in the future.

maxr96 replied:
You are right about documenting the shared services. I updated the documentation, so it should be easier to understand now. In my opinion, though I don't see a big problem here since if you see the service and what URL it is using, you will navigate to check the server code to see what is happening on the REST controller.

ungaralex commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts:
```diff
@@ -367,6 +371,33 @@ export class TutorExerciseDashboardComponent implements OnInit {
+                this.loadAll();
+            },
+            () => {
+                this.loadAll();
```
Why did you remove the subscription to `onResultModified`? I don't think we need to load everything again the user just closed the modal without actually accepting or rejecting a complaint

maxr96 replied:
You are right. I was testing it without this change and removed it. Added back now

ungaralex commented on src/main/webapp/i18n/de/complaintResponse.json:
```diff
@@ -6,7 +6,8 @@
             "rejectComplaint": "Beschwerde zurückweisen",
-            "rejectComplaintTooltip": "Antwort senden und Beschwerde zurückweisen"
+            "rejectComplaintTooltip": "Antwort senden und Beschwerde zurückweisen",
+            "noText": "Bitte gibst du eine Antwortnachricht ein"
```
```suggestion
            "noText": "Bitt gib eine Antwort auf die Beschwerde ein"
```

ungaralex commented on src/main/webapp/i18n/de/error.json:
```diff
@@ -12,6 +12,7 @@
         "validation": "Validierungsfehler auf dem Server.",
         "resultStringNull": "Fehler! Ein Ergebnis Text ist erforderlich.",
         "scoreNull": "Fehler! Eine Bewertung is erforderlich.",
+        "submissionMissing": "Fehler! Die Einreichung ist nicht mehr mit dem Ergebnis verknüpft. Deine Änderungen wurden abgelehnt.",
```
```suggestion
        "submissionMissing": "Fehler! Die Abgabe ist nicht mehr mit dem Ergebnis verknüpft. Deine Änderungen wurden abgelehnt.",
```

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/SubmissionResource.java:
```diff
@@ -131,20 +136,25 @@ public SubmissionResource(SubmissionRepository submissionRepository, ResultRepos
+            var resultId = submission.get().getResult().getId();
+            complaintResponseRepository.deleteByComplaint_Result_Id(resultId);
+            complaintRepository.deleteByResult_Id(resultId);
+            resultRepository.deleteById(resultId);
```
```
// TODO: move this into a ResultService
```

maxr96 replied:
Moved

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/SubmissionResource.java:
```diff
@@ -131,20 +136,25 @@ public SubmissionResource(SubmissionRepository submissionRepository, ResultRepos
      */
     @DeleteMapping("/submissions/{id}")
     @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    @Transactional
```
Please don't

maxr96 replied:
Right. Didn't see this change

maxr96 commented on src/main/webapp/app/list-of-complaints/list-of-complaints.component.ts:
```diff
@@ -98,6 +104,19 @@ export class ListOfComplaintsComponent implements OnInit {
+                    this.loadComplaints();
+                },
+                () => {
+                    this.loadComplaints();
```
No not really. I added `onResultModified` here as well

maxr96 commented:
> I have one additional question: Is there any reason that we keep using wrapper classes in Java for primitive types? I get that we need `Long` for uninitialized model objects, but for all our API endpoints and method parameters it really makes no sense to me to allow `null` values for IDs when we clearly expect an actual ID instead of `null`. Maybe we can discuss this in the next meeting

Thank you for the detailed review @ungaralex . I addressed all the issues you mentioned. I also think it is better to use primitive types. I changed `Long` to `long` in the code I touched. However, it is a good idea to discuss it in the developers meeting to understand if there can be some complications with this approach. One small thing that I already found is that we can't use primitive type in Generics, so our repositories will still extend `JPARepository` with class `Long`

maxr96 commented:
> Some comments here:

Thank you for the comments @jpbernius . Resolved them.

ungaralex commented:
One last thing: please disconnect the result given to the complaing modal on the tutor dashboard from the result showing on the dashboard itself. Any changes on the result in the modal (s.a. changing the score) are immediately refelcted on the dashboard. Because of that, i can change the score in the modal, then just close the dialog or reject the complaint, but the result on the dashboard will still show my changes from the modal until I reload the page. This will  probably be quite confusing to some tutors.

maxr96 commented:
> One last thing: please disconnect the result given to the complaint modal on the tutor dashboard from the result showing on the dashboard itself. Any changes on the result in the modal (s.a. changing the score) are immediately reflected on the dashboard. Because of that, i can change the score in the modal, then just close the dialog or reject the complaint, but the result on the dashboard will still show my changes from the modal until I reload the page. This will probably be quite confusing to some tutors.

Thank you for this suggestion @ungaralex . Indeed that can be very frustrating if discovered by the user. I changed the behavior by passing deep clone of the result to the dialog. Now the result in the dialog and component from which it was called is not referenced anymore

maxr96 commented:
> 
> 
> Is it supposed to say:
> 
> > Complaints about your assessments:
> > 0
> 
> Even though it is actually 1?
> I have seen comments on this before. Not sure if this is addressed in a follow up?
> 
> <img alt="image" width="2000" src="https://user-images.githubusercontent.com/6382716/69952235-0fa26500-14f7-11ea-9e49-aaaf92dd6a36.png">
> 
> <img alt="image" width="2000" src="https://user-images.githubusercontent.com/6382716/69952253-17faa000-14f7-11ea-9551-a2cf2f327ca4.png">

This issue will be fixed in other PR, since it is not related to my changes. `Complaints about your assessment` always shows only accepted complaints. This is how it was designed in the beginning. I am not sure what was the argument for doing so.

maxr96 commented:
> 
> 
> Here, it says:
> 
> > There are no complaints at this moment.
> 
> <img alt="image" width="2000" src="https://user-images.githubusercontent.com/6382716/69952528-a7a04e80-14f7-11ea-8187-e24f48d3378e.png">
> But there are complaints, they are on my assessment so I cannot work on them.
> 
> I think it is not always clear if complaints are on my assessments, or if complaints are on other tutors assessments and I need to resolve them.
> 
> **On a general note:** This works. I can submit complaints, and work on them, only as a different user than doing the initial result.
> 
> **However:**
> 
>     * This is not very intuitive to use! Especially inconsistencies in numbers make me not trust the dashboard!
> 
>     * I feel tutors will not work on complaints as they simply cannot find them?
> 
>     * I think some problems I face are related to the wording: This is one the one side always talking on complaints on my assessment, however I should not work on those, but rather on complaints on others assessments.
> 
> 
> Further, I think that instructors should be allowed to respond to complaints on their own assessments? Not sure what other think on this.

If the `Complaints` are not on your assessment then they will appear in `Complaints` table. 
In my opinion complaints are quite visible. You can find them on the tutor course dashboard and tutor exercise dashboard under clickable `Complaints about your assessment` (here I would agree that it should reacheable under `Total Complaints` or maybe `Active Complaints`). Also in the Tutor Course dashboard you can see the button near number of submissions is becoming red. 
In general, I totally agree that we need to improve the complaints process. I guess we should discuss this on the next developers meeting and create a separate PR to address all issues regarding complaints in general

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:
```diff
@@ -450,6 +450,16 @@ public void filterSensitiveInformation() {
+     * Check if manual results are allowed for the exercise
+     * @return true if manual results are allowed, false otherwise
+     */
+    public boolean areManualResultsAllowed() {
```
Just for documentation: What happens if you postpone the due date? As I was testing your changes I postponed the due date and afterwards, I was able to assess submissions although the due date did not pass yet and students are still able to submit new solutions. We should either disable changing the due date or disable assessing if the due date changes. I think this should be handled in a separate PR.
Also from the students' side 
![image](https://user-images.githubusercontent.com/50635005/70153210-de6c9500-16ae-11ea-9fe4-9c2da644d42d.png)
the due date is messed up as they can submit new results even though the due date has passed. But as I said this might be out of this PRs scope

madwau commented:
Stephan told me that the person reviewing the complaint needs to be different than the one who created the result. With that, I was able to test the update assessment functionality and it worked as expected. Good work!

Smaller UI things that I noticed:

1. Clicking on "Update assessment" or "Reject complaint" does nothing if the complaint's response is left empty. The buttons should be disabled until they can actually be clicked and there should probably be a form validation error displayed to the user that the complaint response can't be empty.

2. The actions buttons are center aligned in some tables which can be confusing since all the other columns are left aligned. At first I thought maybe there is an element left of the button taking up space but not getting rendered properly.

![OnPaste 20191204-173120](https://user-images.githubusercontent.com/7109225/70161717-d4519300-16bc-11ea-8f1b-80bf86c6d826.png)

krusche commented:
I tested the feature and I could still find additional problems :-( 
I immediately fixed the most important ones. 

The remaining issues are:

1. if you reject a complaint, it will not be updated in the Tutor Exercise Dashboard. The tutor has to refresh the page. Same for answering more feedback requests.
2. If you accept a complaint, the result will not be automatically displayed to the student when the navigates to the exercise details page. The student has to reload the application first. Otherwise the old result is shown. I think we can use the websocket here so that students do not need to refresh the page.
3. For some reason there is an [object%20Module] REST Call between complaint-instruction and complaint component that leads to a 404 when navigating to the exercise details page
4. When clicking on X for complaints and more feedback requests for programming exercises, this often leads to a typescript error
<img width="480" alt="ERROR Error Uncaught (in promise) cancel" src="https://user-images.githubusercontent.com/744067/70166235-aff9b480-16c4-11ea-8596-45fa39079699.png">


I think we can address these issues and the ones mentioned by @madwau and @HanyaElhashemy in a follow up PR and now merge this PR because our users want this feature urgently and the main functionality works. Please do that as soon as possible!
